curp/src/server/cmd_worker.rs

```diff
@@ -0,0 +1,537 @@
+            ExeMsg::Execute { cmd, er_tx } => {
+                let er = ce.execute(cmd.as_ref()).await;
+                debug!("cmd {:?} is executed", cmd.id());
+
+                spec.map_lock(|mut spec_l| spec_l.try_remove(cmd.id())); // clean spec pool
+                let _ignore = er_tx.send(er.clone()); // it's ok to ignore the result here because sometimes the result is not needed
+
+                // insert er to cmd_board and check if after sync is already needed
+                let index = {
+                    let mut board_w = cmd_board.write();
+                    let er_ok = er.is_ok();
+                    board_w.insert_er(cmd.id(), er);
+
+                    if !er_ok {
+                        // execution failed, no need to do as
+                        let _ig = board_w.needs_as.remove(cmd.id());
+                        done.notify();
+                        continue;
+                    }
+
+                    if let Some(index) = board_w.needs_as.remove(cmd.id()) {
+                        index
+                    } else {
+                        // mark a cmd done only when after sync is called to ensure as and exe order of conflicted commands
+                        assert!(
+                            board_w
+                                .done_notifiers
+                                .insert(cmd.id().clone(), done)
+                                .is_none(),
+                            "done_notifier should not be inserted twice"
+                        );
+                        continue;
+                    }
+                };
+
+                let asr = ce.after_sync(cmd.as_ref(), index).await;
+                cmd_board.write().insert_asr(cmd.id(), asr);
+                debug!("cmd {:?} after sync is called", cmd.id());
+            }
+            ExeMsg::Reset => {
+                ce.reset().await;
+                debug!("command executor has been reset");
+            }
+        }
+        done.notify();
+    }
+    error!("execute worker stopped unexpectedly");
+}
+
+/// Worker that do after sync
+pub(super) async fn after_sync_worker<C: Command + 'static, CE: 'static + CommandExecutor<C>>(
+    dispatch_rx: CmdAsReceiver<C>,
+    cmd_board: CmdBoardRef<C>,
+    ce: Arc<CE>,
+) {
+    while let Ok((AsMsg { cmd, index }, done)) = dispatch_rx.recv().await {
+        /// Different conditions in after sync
+        enum Condition {
+            /// Call both exe and after sync
+            ExeAndAfterSync,
+            /// Call after sync only
+            AfterSync(DoneNotifier),
+            /// Do nothing
+            Nothing,
+        }
+        let condition = cmd_board.map_write(|mut board_w| {
+            if board_w.needs_exe.remove(cmd.id()) {
+                // The cmd needs both execution and after sync
+                Condition::ExeAndAfterSync
+            } else if let Some(er) = board_w.er_buffer.get(cmd.id()) {
+                debug!("got it");
```

Comment:
got what? The debug message should be more meaningful

Reply:
Fixed. (Forget to remove the debug output.)